sql/expression/case.go

```diff
@@ -52,6 +52,12 @@ func combinedCaseBranchType(left, right sql.Type) sql.Type {
 	if right == types.Null {
 		return left
 	}
+	if types.IsEnum(left) && types.IsEnum(right) {
```

Comment:
Probably worth adding a comment to explain what we're doing here. This seems pretty difficult to reverse engineer and figure out in a year or two when someone new looks at this code and tries to figure out why these cases are in here.